test/common/upstream/hds_test.cc

```diff
@@ -678,5 +687,254 @@ TEST_F(HdsTest, TestSendResponseOneEndpointTimeout) {
+  createHdsDelegate();
+
+  // Create Message with two different health checkers.
+  message.reset(createSimpleMessage());
+  auto new_hc = message->mutable_cluster_health_checks(0)->add_health_checks();
+  new_hc->MergeFrom(message->mutable_cluster_health_checks(0)->health_checks(0));
+  new_hc->mutable_http_health_check()->set_path("/different_path");
+
+  // Create a new active connection on request, setting its status to connected
+  // to mock a found endpoint.
+  EXPECT_CALL(dispatcher_, createClientConnection_(_, _, _, _))
+      .WillRepeatedly(Invoke(
+          [](Network::Address::InstanceConstSharedPtr, Network::Address::InstanceConstSharedPtr,
+             Network::TransportSocketPtr&, const Network::ConnectionSocket::OptionsSharedPtr&) {
+            Network::MockClientConnection* connection =
+                new NiceMock<Network::MockClientConnection>();
+
+            // pretend our endpoint was connected to.
+            connection->raiseEvent(Network::ConnectionEvent::Connected);
+
+            // return this new, connected endpoint.
+            return connection;
+          }));
+
+  EXPECT_CALL(*server_response_timer_, enableTimer(_, _)).Times(AtLeast(1));
+  EXPECT_CALL(async_stream_, sendMessageRaw_(_, false));
+  EXPECT_CALL(test_factory_, createClusterInfo(_)).WillRepeatedly(Return(cluster_info_));
+  EXPECT_CALL(dispatcher_, deferredDelete_(_)).Times(AtLeast(1));
+  // Process message
+  hds_delegate_->onReceiveMessage(std::move(message));
+  hds_delegate_->sendResponse();
+
+  // Save list of health checkers for use later.
+  auto original_hcs = hds_delegate_->hdsClusters()[0]->healthCheckers();
+  ASSERT_EQ(original_hcs.size(), 2);
+
+  // Create a new specifier, but make the second health checker different and add a third.
+  // Then reverse the order so the first one is at the end, testing the hashing works as expected.
+  message.reset(createSimpleMessage());
+  auto new_hc0 = message->mutable_cluster_health_checks(0)->add_health_checks();
+  new_hc0->MergeFrom(message->mutable_cluster_health_checks(0)->health_checks(0));
+  new_hc0->mutable_http_health_check()->set_path("/path0");
+  auto new_hc1 = message->mutable_cluster_health_checks(0)->add_health_checks();
+  new_hc1->MergeFrom(message->mutable_cluster_health_checks(0)->health_checks(0));
+  new_hc1->mutable_http_health_check()->set_path("/path1");
+  message->mutable_cluster_health_checks(0)->mutable_health_checks()->SwapElements(0, 2);
+  hds_delegate_->onReceiveMessage(std::move(message));
+
+  // Get the new health check list from HDS.
+  auto new_hcs = hds_delegate_->hdsClusters()[0]->healthCheckers();
+  ASSERT_EQ(new_hcs.size(), 3);
+
+  // Make sure our first hc from the original list is the same as the third in the new list.
+  EXPECT_EQ(original_hcs[0], new_hcs[2]);
+  EXPECT_TRUE(original_hcs[1] != new_hcs[1]);
+
+  // Check to see that HDS got two requests, and updated two times with it.
+  checkHdsCounters(2, 0, 0, 2);
+}
+
```

Comment:
Great tests for cluster add/remove/update. Do you think it would make sense to add an `hds_integration_test` that verifies a representative update containing an add, remove and update?

Also, I think there should be specific tests for each level/type of hashing, from transport socket to whole message. Is there coverage missing of this ^^?

Reply:
Thanks, I can add an integration test. 

For whole message coverage hash coverage, I updated the existing high-level test that checks for repeated protos to also check for an update, which means that test now has a check for matching hashes and not-matching hashes.

For transport socket, since this is new I had forgotten I needed to add a test for this. I will add a test for this.

Reply:
FYI, I still need to add the implementation test. Will let you know when I am finished with that.